velox/exec/ExchangeClient.cpp

```diff
@@ -47,19 +46,22 @@ void ExchangeClient::addRemoteTaskId(const std::string& taskId) {
       toClose = std::move(source);
     } else {
       sources_.push_back(source);
-      queue_->addSourceLocked();
       // Put new source into 'producingSources_' queue to prioritise fetching
       // from these to find out whether these are productive or not.
       producingSources_.push(source);
-
-      requestSpec = pickSourcesToRequestLocked();
     }
   }
 
   // Outside of lock.
   if (toClose) {
     toClose->close();
   } else {
+    queue_->addSourceLocked();
```

Comment:
This must be called while holding queue's lock. I don't see this being the case. Would you double check?